pkg/webhook/server.go

```diff
@@ -74,14 +92,67 @@ func (s *Server) setDefaults() {
 	s.webhooks = map[string]http.Handler{}
 	s.webhookMux = http.NewServeMux()
 
+	// TODO: probably allow set Port by env.
+	p, err := strconv.Atoi(os.Getenv("WEBHOOKPORT"))
+	if err != nil {
+		log.Error(err, "can't convert WEBHOOKPORT environment variable to int")
+	} else {
+		s.Port = p
+	}
 	if s.Port <= 0 {
 		s.Port = 443
 	}
+
 	if len(s.CertDir) == 0 {
 		s.CertDir = path.Join("/tmp", "k8s-webhook-server", "serving-certs")
 	}
 }
 
+// EnableWebhooksFor registers mutating (defaulting) and validation webhooks for the provided types that
+// implement Defaulter and (or) Validator. Passing in core types doesn't work, since the core types implement
+// neither the Defaulter nor the Validator interfaces.
+// It returns a NotImplementDefaulterValidatorInterfacesError error if a provided object implements
+// neither the Defaulter interface nor the Validator interface.
+func (s *Server) EnableWebhooksFor(objects ...runtime.Object) error {
+	if s.Scheme == nil {
+		return fmt.Errorf("field Scheme must be set or injected before invoking EnableWebhooksFor")
+	}
+
+	// Create a webhook for each type
+	for _, object := range objects {
+		gvk, err := apiutil.GVKForObject(object, s.Scheme)
+		if err != nil {
+			return nil
+		}
+
+		mwh := newDefaultingWebhookFor(object)
```

Comment:
we should opportunistically merge validating and defaulting hooks into a single hook when possible, to avoid the extra round-trip

Reply:
> we should opportunistically merge validating and defaulting hooks into a single hook when possible, to avoid the extra round-trip

This means you want to put defaulting and validating hooks in one mutatingWebhook. But doing it this way is not safe, since there may be another mutating webhook changes the object after your validation logic. The safe way to do it is making validation logic in a validating webhook, which is always run after all mutating webhooks.

Reply:
good point :+1: